### ISSUE TITLE
Including exogenous variables in SARIMAX. Probably an easy solution.

### ISSUE THREAD

**Issue by rosato11** [statsmodels/statsmodels#4284]
I want to include an exog variable in my model which is mean temp.  My code is below.

train = data.loc[:'2012-12-13','age6-15']
exog = data.loc[:'2012-12-13','Daily mean temp']
mod = sm.tsa.statespace.SARIMAX(train, exog=exog, trend='n', order=(0,1,0), seasonal_order=(1,1,1,52))
results = mod.fit()
predictions = results.predict(start = '2012-12-13', end = '2016-12-22', dynamic= True)

I am getting this error. Thanks.

ValueError: Out-of-sample forecasting in a model with a regression component requires additional exogenous values via the `exog` argument.

**Comment by josef-pkt** [comment]
As the error message says: you need to provide an exog in `predict` for out-of-sample forecasting.

It is not possible to forecast without knowing all the explanatory variables for the forecast periods.

BTW: AFAICS, you are not including a constant

**Comment by rosato11** [comment]
I have a dataset of weekly rotavirus count from 2004 - 2016. A vaccine was introduced in 2013.
I have been able to make a prediction for 2013 - 2014 by training the model with the data from 2004 - 2013. I can then look at the predicted vs the actual when the vaccine was introduced. 

I have temperature data from 2004 - 2016. Am I right by assuming that I can not use the full temp data (2004-2016) to make predictions for rotavirus during 2013-2016 because the endog and exog variables need to be of the same size?

**Comment by rosato11** [comment]
when I change the exog to the size of my temp data (seen below)
train = data.loc[:'2012-12-13','age6-15']
exog = data.loc[:'2016-12-22','Daily mean temp']

i get the error: ValueError: The indices for endog and exog are not aligned

Can I not use the temp data to help predict the years for rotavirus count between: 2013-2016?

**Comment by josef-pkt** [comment]
you need to keep the exog in the training/estimation sample the same length (and periods/index) as your endog. 
However, you need to specify a new exog in predict, i.e.

estimate as before

```
train = data.loc[:'2012-12-13','age6-15']
exog = data.loc[:'2012-12-13','Daily mean temp']
mod = sm.tsa.statespace.SARIMAX(train, exog=exog, trend='n', order=(0,1,0), seasonal_order=(1,1,1,52))
results = mod.fit()
```

then define and use the forecast exog for predict
```
exog_forecast = data.loc['2012-12-13':'2016-12-22',,'Daily mean temp']
predictions = results.predict(start = '2012-12-13', end = '2016-12-22' exog=exog_forecast, dynamic= True)
```

Note: There was an ambiguity in earlier version about whether exog in `predict` includes the full exog (train plus forecast sample) or just the forecast/predict sample. I'm not sure how SARIMAX is handling this now.

**Comment by rosato11** [comment]
Thank you very much for the reply.
I now get the error: 
ValueError: Provided exogenous values are not of the appropriate shape. Required (210, 1), got (211L,).

Is that referring to the same as this? [https://github.com/statsmodels/statsmodels/issues/3907](url)

That the exog values need to be in a 2 dimensional dataframe?

**Comment by josef-pkt** [comment]
> Required (210, 1), got (211L,).

my guess its that you need to start the exog at the first out-of-sample observation,
i.e. '2012-12-13' is in the training/estimation sample (assuming pandas includes the endpoint in the time slice)
and keep exog_forecast as a dataframe to avoid #3907
`exog_forecast = data.loc['2012-12-14':'2016-12-22',['Daily mean temp']]`

Check if that produces a correct looking forecast.

**Comment by rosato11** [comment]
Once again thanks for the reply.
I am now getting the error:
ValueError: Provided exogenous values are not of the appropriate shape. Required (208, 1), got (208L,).

Is this similar to #3907  that I need to make it a data frame before the prediction?

**Comment by rosato11** [comment]
Got it working. It needed to be a 2 dimensional dataframe!
Thanks for all your help.

**Comment by Santosh6773** [comment]
@rosato11 
It needed to be a 2 dimensional dataframe! >> Can you please share at which point you applied the fix?
Thanks a lot ! 👍

**Comment by RayenFekih** [comment]
> @rosato11
> It needed to be a 2 dimensional dataframe! >> Can you please share at which point you applied the fix?
> Thanks a lot ! 👍
in his case he needs to add [-208:,None] to make sure the shape is right so he writes:
 exog_forecast = data.loc['2012-12-14':'2016-12-22',['Daily mean temp']][-208:,None]

**Comment by MchlUh** [comment]
@Santosh6773 Please share the way you fixed this 🙏 🙏 🙏

**Comment by bashtage** [comment]
Going to close this has he since it was answered and there are no actions for devs.  Please continue to clarify the solution so other will benefit.

**Comment by ChadFulton** [comment]
The underlying problem in the later comments of this issue was that the `exog` passed to `forecast` had to be a DataFrame, even if it was a single column. This problem has been fixed in v0.12, so I suggest that you update Statsmodels.

If there is still a problem with passing `exog` to `forecast` or `predict`, please open a new issue with a description of what is happening.
